src/pages/BookingDetails/BookingDetails.tsx

```diff
@@ -0,0 +1,463 @@
+import axios from "axios";
+import React, { useEffect, useRef, useState } from "react";
+import { Button, Container, Form } from "react-bootstrap";
+import { useLocation, useNavigate } from "react-router-dom";
+import { toast } from "react-toastify";
+import { UseFetch } from "../../customHook/UseFetch";
+import { Addon } from "../../types/types";
+import "./bookingdetails.css";
+
+interface ResAddon {
+  data: Addon[];
+  loading: boolean;
+}
+
+interface State {
+  no: number;
+  checkin: Date;
+  checkout: Date;
+  roomType: string;
+  roomPrice: number;
+  totalPrice: number;
+  key: string;
+}
+
+export const BookingDetails = () => {
+  const location = useLocation();
+  const { no, checkin, checkout, roomType, roomPrice, totalPrice, key }: State =
+    location.state;
+  console.log(location.state);
+  const nameRef = useRef<HTMLInputElement>(null);
+  const emailRef = useRef<HTMLInputElement>(null);
+  const specialReqRef = useRef<HTMLTextAreaElement>(null);
+  const couponRef = useRef<HTMLInputElement>(null);
+
+  const [coupon, setCoupon] = useState("");
+
+  const buttonRef = useRef<HTMLButtonElement>(null);
+  const buttonRemRef = useRef<HTMLButtonElement>(null);
+  const checkRef = useRef<HTMLInputElement>(null);
+
+  const [apply, setApply] = useState(false);
+
+  const navigate = useNavigate();
+
+  let total_price = totalPrice;
+
+  const { data, loading }: ResAddon = UseFetch(`addon`);
+  let addOns = {};
+  if (!loading) {
+    data.forEach((item) => {
+      addOns[item.name] = Number(item.price);
+    });
+  }
+
+  const [selectCheck, setSelectCheck] = useState({
+    addon: [],
+  });
+
+  const [total, setTotal] = useState(0);
+  const [discount, setDiscount] = useState(0);
+
+  useEffect(() => {
+    Object.values(filtAddOn()).forEach((each: any) => {
+      total_price += each;
+    });
+    setTotal(
+      total_price +
+        addOnPrice() -
+        (total_price + addOnPrice()) * (discount / 100)
+    );
+  }, [selectCheck]);
+
+  const handleAddon = (e) => {
+    if (e.target.checked) {
+      setSelectCheck({
+        addon: [...selectCheck.addon, e.target.value],
+      });
+    }
+    if (!e.target.checked) {
+      setSelectCheck({
+        addon: selectCheck.addon.filter((item) => item !== e.target.value),
+      });
+    }
+  };
+
+  const filtAddOn = () => {
+    let a = {};
+    selectCheck.addon.forEach((each) => {
+      if (Object.keys(addOns).includes(each)) {
+        a[each] = addOns[each];
+      }
+    });
+    return a;
+  };
+
+  const handleCoupon = async (e) => {
+    e.preventDefault();
+    // const coupon = couponRef.current?.value;
+    const res = await axios.post("coupon/validate", {
```

Comment:
move this to a separate function `validateCoupon` and then call the `applyCoupon `method